flutter/lib/benchmark/state.dart

```diff
@@ -392,6 +392,12 @@ class BenchmarkState extends ChangeNotifier {
       performanceRunInfo.loadgenInfo!;
 
       final performanceResult = performanceRunInfo.result;
+      if (!performanceResult.accuracyNormalized.isFinite ||
+          performanceResult.accuracyNormalized > 1.0 ||
+          !performanceResult.accuracyNormalized2.isFinite ||
+          performanceResult.accuracyNormalized2 > 1.0) {
```

Comment:
We use negative values to indicate that accuracy can not be calculated. For example, if groundtruth files are missing.

Reply:
Maybe better to make the `accuracyNormalized` optional? It should be `null` if the there is no value.
Here we check for programming error with `.isFinite` and logic error with `> 1.0` but another case of logic error with `< 0` is not caught.

Reply:
Case of `< 0` is not an error. It is a perfectly valid value for accuracy in the [RunResult class](https://github.com/mlcommons/mobile_app_open/blob/ab6af7572e5f975da354400cabfc15449d81f17d/flutter/lib/backend/bridge/run_result.dart).
Both accuracy values are already optional in the `BenchmarkRunResult` class.
https://github.com/mlcommons/mobile_app_open/blob/ab6af7572e5f975da354400cabfc15449d81f17d/flutter_common/lib/data/results/benchmark_result.dart#L65
We have a check for this few lines below:
https://github.com/mlcommons/mobile_app_open/blob/ab6af7572e5f975da354400cabfc15449d81f17d/flutter/lib/benchmark/state.dart#L455

`RunResult` is used to transfer data from native code to business logic.
I think it would be better to process it in one place.

Reply:
In that case, we should only check here if `accuracyNormalized` is a valid `double` number. The check `> 1.0` should be moved to the same place with the check `< 0.0 `

Reply:
What would we do there if we encounter value `> 1.0`?
Negative values is a special case because we have a convention to use it for when groundtruth is missing.
We don't have any similar convention for values `> 1.0`. The only thing we can do about it is show an error, and this is what we do here. It wouldn't make sense to move the check few lines below, and then show the same error.

Reply:
The check `> 1.0` is a logic check, the same as `< 0` so it should be in one place. 
That helps others when reading the code understand why we do the check.

Reply:
I don't see how `NaN`/`infinity` check is different from `> 1.0` check. All these values are just invalid values. They can only occur if our native code is broken.
`< 0.0` is different because it's a perfectly valid value, we expect it.

Reply:
>The check `>1.0` is a logic check

Just to clarify my point: it isn't a logic check.

Reply:
> > The check `>1.0` is a logic check
> 
> Just to clarify my point: it isn't a logic check.

Yes, it is. In normal programming, we don't expect a number to be `NaN` or `infinity`, that is a programming error.
In other side, any check for its specific value is a logic check because in this context of this app, we expect the number to be in a certain range to be valid. Maybe later in the future, the logic is changed to have normalized accuracy in range `[0..100]` instead of `[0..1]`.
It may add some more lines of code to have those check separated but it makes clear what is going on. And in an opensources project with multiple developers working on it, it's important that other developers understand what is going on without too much effort.

To be concrete, I would write the check like this:
```dart


final performanceResult = performanceRunInfo.result;
if (!performanceResult.accuracyNormalized.isFinite ||
    !performanceResult.accuracyNormalized2.isFinite) {
  throw '${benchmark.info.taskName}: performance run: accuracy is not a finite number (backend may be corrupted)';
}

Accuracy? perfModeAccuracy;
// normalized accuracy is expected to be in range [0..1]
if (performanceResult.accuracyNormalized >= 0 &&
    performanceResult.accuracyNormalized <= 1) {
  perfModeAccuracy = Accuracy(
    normalized: performanceResult.accuracyNormalized,
    formatted: performanceResult.accuracyFormatted,
  );
}

Accuracy? perfModeAccuracy2;
// normalized accuracy is expected to be in range [0..1]
if (performanceResult.accuracyNormalized2 >= 0 &&
    performanceResult.accuracyNormalized2 <= 1) {
  perfModeAccuracy2 = Accuracy(
    normalized: performanceResult.accuracyNormalized2,
    formatted: performanceResult.accuracyFormatted2,
  );
}

benchmark.performanceModeResult = BenchmarkResult(
  throughput: performanceRunInfo.throughput,
  accuracy: perfModeAccuracy,
  accuracy2: perfModeAccuracy2,
  backendName: performanceResult.backendName,
  acceleratorName: performanceResult.acceleratorName,
  batchSize: benchmark.benchmarkSettings.batchSize,
  validity: performanceRunInfo.loadgenInfo!.validity,
);

```

Reply:
Ok, I'll create a separate variable to indicate that accuracy is invalid, instead of using negative values.